Modifying the number of leave days calculation part.

meatechsupport opened the pull request:
we don't need to exclude the Holiday list (that comes in between) from the total number of leaves applied.

meatechsupport committed:
Modifying the number of leave days calculation part.

we don't need to exclude the Holiday list (that comes in between) from the total number of leaves applied.

Add option in the leave type

added a field in the leave type to include and exclude the holidays from
the tolal leave applied days

Added the field

Added a field Include Holiday to leave Type doctype

changed the lable

changed the lable from "Include Holiday" to "Include holidays within leaves
as leaves"

Rearranged the function

moved holidays = leave_app.get_holidays() under if
Corrected 'total_leave_days' : flt(tot_days)-flt(holidays)

Adding test case

added the test case

Added test case

Added test case to test_leave_application.py

adding default value

added default value and corrected the syntax.

IndentationError

removed extra tabs after

rmehta commented:
@meatechsupport can you create a new field in HR Settings like "Holidays between leaves are counted as leaves" (`include_holidays_in_leaves`) and then base this calculation based on what the setting is?

you can use `frappe.db.get_value("HR Settings", "HR Settings", "include_holidays_in_leaves")` to check the value of the setting.

rmehta commented:
@meatechsupport do you still want to do this. We can do it as paid development, in 1 hr? https://erpnext.com/pricing/paid-feature-development

meatechsupport commented:
@rmehta , sorry for the delayed input. i believe the adding the check-box in the "Leave Type" doctype will be more friendly than the  HR settings. Currently i have added Check-box "Include Holiday" in the Leave Type and modified the leave_application.py accordingly.

rmehta commented:
@meatechsupport Better. You have to also commit the leave_type.json file with the new field.

meatechsupport commented:
@rmehta committed the leave_type.json.

rmehta commented:
@meatechsupport thanks, can you change the label to "Include holidays within leaves as leaves" so that it is explicit?

meatechsupport commented:
@rmehta thanks, the label has been changed accordingly.

nabinhait commented on erpnext/hr/doctype/leave_application/leave_application.py:
```diff
@@ -239,9 +239,14 @@ def get_total_leave_days(leave_app):
-		}
+			'total_leave_days' : flt(tot_days) - flt(holidays)
+			}
+			else:
```
Indentation is not correct

nabinhait commented on erpnext/hr/doctype/leave_application/leave_application.py:
```diff
@@ -239,9 +239,14 @@ def get_total_leave_days(leave_app):
 	if not leave_app.half_day:
 		tot_days = date_diff(leave_app.to_date, leave_app.from_date) + 1
 		holidays = leave_app.get_holidays()
+		if frappe.db.get_value("Leave Type", self.leave_type, "include_holiday"):
```
`holidays = leave_app.get_holidays()` can be inside if condition

meatechsupport commented:
@nabinhait, i hope the changes are good enough to go. :)

rmehta commented:
@meatechsupport can you write a simple test case? Just copy this one and add your condition in a new Leave Type:

https://github.com/frappe/erpnext/blob/develop/erpnext/hr/doctype/salary_slip/test_salary_slip.py#L28-L29

That will make is future-proof

meatechsupport commented:
hi @rmehta, test case in the sense!! can please explain?

rmehta commented:
Did you see the link? A test case ensures that your code runs as it is designed to without anyone checking and if a future update breaks it, then it will throw a failed test.

meatechsupport commented:
hi @rmehta, have copied it to the test file.

meatechsupport commented:
@rmehta , any update

meatechsupport commented:
@rmehta i have tested it on the virtual box image and it was working fine. so can pleas tell what need to be done. sorry still didn't get clear picture about adding test case here. :(

rmehta commented:
@meatechsupport did you read the other test cases in the file?

Broad introduction to unit testing

https://en.wikipedia.org/wiki/Unit_testing

meatechsupport commented:
@rmehta  please let me know if further changes required. :)

rmehta commented:
Nice! Now cleanup.

1. Reset the value of include_holidays in leave_type after the test has passed, because the tests are run in random order and you should leave any changes as they were before the test.
2. Squash your 7 commits into one commit so that it looks cleaner (Google how to do this)
3. Rebase your branch

meatechsupport commented:
@rmehta how can we know if the status of the test, if it passed or failed?
should i wait till the test finish to squashing and rebasing

rmehta commented:
No just reset the leave_type setting using 

```
frappe.db.set_value("Leave Type", self.leave_type, "include_holiday", [original setting])
```

and squash and rebase.

To test locally, check this:

```
bench --site [your site] run-tests --doctype "Leave Application"
```

meatechsupport commented:
@rmehta i really appreciate your help. 

could you please specify which file to edit, leave_type.py or leave_type.json?

rmehta commented:
Your test case. At the end of your test, reset the leave type property as it may affect other tests.

meatechsupport commented:
@rmehta squashed and rebased.

meatechsupport committed:
IndentationError

alligned code

meatechsupport committed:
Indentation Error

realligned

meatechsupport committed:
Edited the Test case

meatechsupport committed:
removed testcase

meatechsupport commented:
@rmehta even if i remove the Test Case it fails in testing other function. please let me know if i ignore and add the test case again. waiting  for your response.

rmehta commented:
@meatechsupport @nabinhait  will merge your PR later today.

nabinhait commented:
Merged. Made some related changes as well, https://github.com/frappe/erpnext/commit/b24d2efc4b99239f935b87a685e46cb5da2497c6.
